### PR TITLE
Feature/refactor checks

### PR DESCRIPTION
**Pull request by rocktavious**
## Issues

https://github.com/OpsLevel/team-platform/issues/309

## Changelog

[check_manual public docs](https://registry.terraform.io/providers/OpsLevel/opslevel/latest/docs/resources/check_manual)

- [ ] List your changes here
- [ ] Make a `changie` entry

### Using this file:
```tf
# main.tf
resource "opslevel_check_manual" "example" {
  name      = "foo"
  enable_on = "2024-05-23T14:14:18.782000Z"
  category  = data.opslevel_rubric_category.security.id
  level     = data.opslevel_rubric_level.bronze.id
  owner     = data.opslevel_team.devs.id
  filter    = data.opslevel_filter.tier1.id
  update_frequency = {
    starting_date = "2024-05-23T14:14:18Z"
    time_scale    = "week"
    value         = 1
  }
  update_requires_comment = false
  notes                   = "Optional additional info on why this check is run or how to fix it"
}

resource "time_static" "initial" {}
```

### Create Check Manual, `terraform apply`
```tf
  # opslevel_check_manual.example will be created
  + resource "opslevel_check_manual" "example" {
      + category                = "Z2lkOi8vb3BzbGV2ZWwvQ2F0ZWdvcnkvOTY1"
      + description             = (known after apply)
      + enable_on               = "2024-05-23T14:14:18.782000Z"
      + enabled                 = (known after apply)
      + filter                  = "Z2lkOi8vb3BzbGV2ZWwvRmlsdGVyLzQzNjk"
      + id                      = (known after apply)
      + last_updated            = (known after apply)
      + level                   = "Z2lkOi8vb3BzbGV2ZWwvTGV2ZWwvNTIw"
      + name                    = "foo"
      + notes                   = "Optional additional info on why this check is run or how to fix it"
      + owner                   = "Z2lkOi8vb3BzbGV2ZWwvVGVhbS8xMTcwNQ"
      + update_frequency        = {
          + starting_date = "2024-05-23T14:14:18Z"
          + time_scale    = "week"
          + value         = 1
        }
      + update_requires_comment = false
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

### Update this file:
```tf
resource "opslevel_check_manual" "example" {
  name      = "foo"
  enable_on = "2024-05-23T14:14:18.782000Z"
  category  = data.opslevel_rubric_category.security.id
  level     = data.opslevel_rubric_level.bronze.id
  owner     = data.opslevel_team.devs.id
  filter    = data.opslevel_filter.tier1.id
  update_frequency = {
    starting_date = "2024-05-23T14:14:18Z"
    time_scale    = "day"
    value         = 1
  }
  update_requires_comment = false
  notes                   = "Optional additional info on why this check is run or how to fix it"
}
```

### Update Check Manual, `terraform apply`
```tf
  ~ resource "opslevel_check_manual" "example" {
      + description             = (known after apply)
      ~ enabled                 = false -> (known after apply)
        id                      = "Z2lkOi8vb3BzbGV2ZWwvQ2hlY2tzOjpNYW51YWwvMjQwNDA"
      ~ last_updated            = "Wednesday, 10-Apr-24 13:44:49 CDT" -> (known after apply)
        name                    = "foo"
      ~ update_frequency        = {
          ~ time_scale    = "week" -> "day"
            # (2 unchanged attributes hidden)
        }
        # (7 unchanged attributes hidden)
    }

```

### Destroy Check Manual, `terraform apply`
```tf
  - resource "opslevel_check_manual" "example" {
      - category                = "Z2lkOi8vb3BzbGV2ZWwvQ2F0ZWdvcnkvOTY1" -> null
      - enable_on               = "2024-05-23T14:14:18.782000Z" -> null
      - enabled                 = false -> null
      - filter                  = "Z2lkOi8vb3BzbGV2ZWwvRmlsdGVyLzQzNjk" -> null
      - id                      = "Z2lkOi8vb3BzbGV2ZWwvQ2hlY2tzOjpNYW51YWwvMjQwNDA" -> null
      - last_updated            = "Wednesday, 10-Apr-24 13:45:07 CDT" -> null
      - level                   = "Z2lkOi8vb3BzbGV2ZWwvTGV2ZWwvNTIw" -> null
      - name                    = "foo" -> null
      - notes                   = "Optional additional info on why this check is run or how to fix it" -> null
      - owner                   = "Z2lkOi8vb3BzbGV2ZWwvVGVhbS8xMTcwNQ" -> null
      - update_frequency        = {
          - starting_date = "2024-05-23T14:14:18Z" -> null
          - time_scale    = "week" -> null
          - value         = 1 -> null
        } -> null
      - update_requires_comment = false -> null
    }

```

### CODE REVIEW

File: opslevel/resource_opslevel_check_base.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	"time"
+)
+
+var checkBaseAttributes = map[string]schema.Attribute{
+	"category": schema.StringAttribute{
+		Description: "The id of the category the check belongs to.",
+		Required:    true,
+		Validators:  []validator.String{IdStringValidator()},
+	},
+	"description": schema.StringAttribute{
```
Comment (davidbloss) [inline opslevel/resource_opslevel_check_base.go:22]: I noticed `description` isn't part of the [check_manual](https://registry.terraform.io/providers/OpsLevel/opslevel/latest/docs/resources/check_manual) schema, or any of the other `opslevel_check_*` resources

File: opslevel/resource_opslevel_check_base.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	},
+	"last_updated": schema.StringAttribute{
+		Optional: true,
+		Computed: true,
+	},
+}
+
+type CheckBaseModel struct {
+	Category    types.String      `tfsdk:"category"`
+	Description types.String      `tfsdk:"description"`
```
Comment (davidbloss) [inline opslevel/resource_opslevel_check_base.go:75]: Similar to note above. Looks like `description` should be removed?

File: opslevel/resource_opslevel_check_base.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+		Notes:    model.Notes.ValueStringPointer(),
+		Owner:    opslevel.NewID(model.Owner.ValueString()),
+	}
+	return opslevel.NewCheckCreateInputTypeOf[T](input), diags
+}
+
+func NewCheckUpdateInputFrom[T any](model CheckBaseModel) (*T, diag.Diagnostics) {
+	enabledOn, diags := asISO8601(model.EnableOn)
+	input := opslevel.CheckUpdateInput{
+		Category: *opslevel.NewID(model.Category.ValueString()),
```
Comment (davidbloss) [inline opslevel/resource_opslevel_check_base.go:112]:
```suggestion
		Category: opslevel.ID(model.Category.ValueString()),
```

File: opslevel/resource_opslevel_check_base.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+		Enabled:  model.Enabled.ValueBoolPointer(),
+		EnableOn: enabledOn,
+		Filter:   opslevel.NewID(model.Filter.ValueString()),
+		Level:    *opslevel.NewID(model.Level.ValueString()),
+		Id:       *opslevel.NewID(model.Id.ValueString()),
```
Comment (davidbloss) [inline opslevel/resource_opslevel_check_base.go:117]:
```suggestion
		Level:    opslevel.ID(model.Level.ValueString()),
		Id:       opslevel.ID(model.Id.ValueString()),
```

File: opslevel/resource_opslevel_check_base.go (hunk 3 above)
Comment (davidbloss) [reply opslevel/resource_opslevel_check_base.go:112]: I just found `asID()`, that might be better

File: opslevel/resource_opslevel_check_manual.go [hunk 5]
```diff
@@ -1,148 +1,205 @@
...
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &model)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	input, diags := NewCheckUpdateInputFrom[opslevel.CheckManualUpdateInput](model.CheckBaseModel)
+	resp.Diagnostics.Append(diags...)
+	input.UpdateRequiresComment = model.UpdateRequiresComment.ValueBoolPointer()
+	// TODO: this is fucking ugly
```
Comment (davidbloss) [inline opslevel/resource_opslevel_check_manual.go:162]: Language, sir 🤭

File: opslevel/terraform_type_conversions.go [hunk 6]
```diff
@@ -47,3 +50,14 @@ func unquote(value string) string {
 	}
 	return value
 }
+
+// asID converts a types.String to an opslevel.ID
+func asID(input types.String) opslevel.ID {
```
Comment (davidbloss) [inline opslevel/terraform_type_conversions.go:55]: 👏 👏 bravo!

File: tests/resources.tf [hunk 7]
```diff
@@ -66,3 +66,35 @@ resource "opslevel_user" "mock_user_admin" {
...
+    starting_date = "2020-02-12T06:36:13Z"
+    time_scale    = "week"
+    value         = 1
+  }
+  update_requires_comment = false
+  notes                   = "Optional additional info on why this check is run or how to fix it"
+}
+
+# Figure out a way to test assert problem using enabled & enable_on
+#resource "opslevel_check_manual" "error" {
```
Comment (davidbloss) [inline tests/resources.tf:91]: Commented out code here

File: opslevel/resource_opslevel_check_base.go (hunk 1 above)
Comment (rocktavious) [reply opslevel/resource_opslevel_check_base.go:23]: switched to computed

File: opslevel/resource_opslevel_check_base.go (hunk 2 above)
Comment (rocktavious) [reply opslevel/resource_opslevel_check_base.go:76]: switched to computed

File: tests/resources.tf (hunk 7 above)
Comment (rocktavious) [reply tests/resources.tf:91]: this is here to remind us

**Review by davidbloss — APPROVED** [review]
🚢 🚢

**Comment by davidbloss** [comment]
Needs tophatting but looks good

File: opslevel/resource_opslevel_check_manual.go (hunk 5 above)
Comment (rocktavious) [reply opslevel/resource_opslevel_check_manual.go]: removed and cleaned up

**Review by davidbloss — APPROVED** [review]
LGTM! 🎉

**Review by davidbloss — APPROVED** [review]
LGTM 🎉

**Pull request merged**